pennylane/qnodes/base.py

```diff
@@ -474,8 +474,12 @@ def _construct(self, args, kwargs):
         """
         # pylint: disable=attribute-defined-outside-init, too-many-branches, too-many-statements
 
-        if self.arg_vars is None or self.kwarg_vars is None:
```

Comment:
So basically, just remove this line?

Reply:
Yes, that's very true! The reason for reverting the logic is more on the side of performance, see the comment at https://github.com/XanaduAI/pennylane/pull/446/files#r370309305

Reply:
Yep --- @antalszava did some benchmarking, and found that removing the caching (and having `_make_variables` be called with every evaluation) adds a _lot_ of overhead, mostly due to the use of `functools.inspect`.

Reply:
(in one particular case, a QNode evaluation loop was 12s slower with this line removed, I believe?)

Reply:
I double-checked this on a very simple example I had a look at and realized that `timeit.timeit` does not return results averaged over the number of runs. What this means, is that this result was actually in the regime of microseconds (a mistake I made, unfortunately).

This is the small example I looked at (now using the `%timeit` syntax which *does* return valid statistics):
```python
import inspect
def func(a, b, c, d=1234):
    return 2

%timeit inspect.getfullargspec(func)
```

Reply:
So the question would rather be if the flattening part would really introduce overhead. If not, then the logic in this PR should be simplified to removing caching only.

Reply:
If `functools.inspect` is the bottleneck, then we can also just cache the result of it, because this will not be subject to change even if the contents of the variables change

Reply:
Could it perhaps change if we have `**kwargs` in the signature of the qfunc? For example:
```python
@qml.qnode(dev)  
def circuit(a, **kwargs):  
    c = kwargs.get('c', False) 
    qml.RX(a, wires=[0]) 
    if c: 
        qml.RY(a, wires=[0]) 
    return qml.expval(qml.PauliZ(0))  
  
circuit(0.3) 
circuit(0.5, c=True) 
```

Reply:
PennyLane 0.7 results:

30 evaluations (`num_evaluations = 30`):

```python
In [2]: qml.about()                                                                                                                                                                                         
WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
Name: PennyLane
Version: 0.7.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /anaconda3/lib/python3.7/site-packages
Requires: scipy, autograd, toml, semantic-version, numpy, appdirs, networkx
Required-by: PennyLane-SF, PennyLane-Forest, PennyLane-Qchem, PennyLane-PQ, PennyLane-qsharp, PennyLane-Cirq
Platform info:           Linux-4.19.11-041911-generic-x86_64-with-debian-buster-sid
Python version:          3.7.6
Numpy version:           1.16.4
Scipy version:           1.4.1
Installed devices:
- default.gaussian (PennyLane-0.7.0)
- default.qubit (PennyLane-0.7.0)
- expt.tensornet (PennyLane-0.7.0)
- expt.tensornet.tf (PennyLane-0.7.0)
- strawberryfields.fock (PennyLane-SF-0.6.0)
- strawberryfields.gaussian (PennyLane-SF-0.6.0)
- forest.numpy_wavefunction (PennyLane-Forest-0.6.0)
- forest.qpu (PennyLane-Forest-0.6.0)
- forest.qvm (PennyLane-Forest-0.6.0)
- forest.wavefunction (PennyLane-Forest-0.6.0)
- projectq.classical (PennyLane-PQ-0.7.0.dev0)
- projectq.ibm (PennyLane-PQ-0.7.0.dev0)
- projectq.simulator (PennyLane-PQ-0.7.0.dev0)
- microsoft.QuantumSimulator (PennyLane-qsharp-0.6.0)
- cirq.simulator (PennyLane-Cirq-0.8.0)

In [5]: import pennylane as qml  
   ...: import numpy as np  
   ...:   
   ...: dev = qml.device('default.qubit', wires=1)  
   ...:   
   ...: @qml.qnode(dev)  
   ...: def circuit(a, b=1):  
   ...:     for idx in range(b):  
   ...:         qml.RX(a[idx], wires=[0])  
   ...:     return qml.expval(qml.PauliX(0))  
   ...:   
   ...: def func():  
   ...:     num_evaluations = 30  
   ...:     for i in range(num_evaluations):  
   ...:         params = np.random.rand(i)  
   ...:         circuit(params, b=i)  
   ...: %timeit func()                                                                                                                                                                                      
4.32 ms ± 34.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

100 evaluations (`num_evaluations = 100`):

```python
In [6]: import pennylane as qml  
   ...: import numpy as np  
   ...:   
   ...: dev = qml.device('default.qubit', wires=1)  
   ...:   
   ...: @qml.qnode(dev)  
   ...: def circuit(a, b=1):  
   ...:     for idx in range(b):  
   ...:         qml.RX(a[idx], wires=[0])  
   ...:     return qml.expval(qml.PauliX(0))  
   ...:   
   ...: def func():  
   ...:     num_evaluations = 100 
   ...:     for i in range(num_evaluations):  
   ...:         params = np.random.rand(i)  
   ...:         circuit(params, b=i)  
   ...: %timeit func()                                                                                                                                                                                      
27.2 ms ± 3.43 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Here going from 30 to 100 evaluations yielded 6.2962962 times of increase in execution time.

--------------------------------

PennyLane 0.9.0-dev with no caching (simply removing the caching condition)

```python
In [10]: qml.about()                                                                                                                                                                                        
WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
Name: PennyLane
Version: 0.9.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /xanadu/pennylane
Requires: numpy, scipy, networkx, autograd, toml, appdirs, semantic-version
Required-by: PennyLane-SF, PennyLane-Forest, PennyLane-Qchem, PennyLane-PQ, PennyLane-qsharp, PennyLane-Cirq
Platform info:           Linux-4.19.11-041911-generic-x86_64-with-debian-buster-sid
Python version:          3.7.6
Numpy version:           1.16.4
Scipy version:           1.4.1
Installed devices:
- default.gaussian (PennyLane-0.9.0.dev0)
- default.qubit (PennyLane-0.9.0.dev0)
- default.tensor (PennyLane-0.9.0.dev0)
- default.tensor.tf (PennyLane-0.9.0.dev0)
- strawberryfields.fock (PennyLane-SF-0.6.0)
- strawberryfields.gaussian (PennyLane-SF-0.6.0)
- forest.numpy_wavefunction (PennyLane-Forest-0.6.0)
- forest.qpu (PennyLane-Forest-0.6.0)
- forest.qvm (PennyLane-Forest-0.6.0)
- forest.wavefunction (PennyLane-Forest-0.6.0)
- projectq.classical (PennyLane-PQ-0.7.0.dev0)
- projectq.ibm (PennyLane-PQ-0.7.0.dev0)
- projectq.simulator (PennyLane-PQ-0.7.0.dev0)
- microsoft.QuantumSimulator (PennyLane-qsharp-0.6.0)
- cirq.simulator (PennyLane-Cirq-0.8.0)

In [11]: import pennylane as qml  
    ...: import numpy as np  
    ...:   
    ...: dev = qml.device('default.qubit', wires=1)  
    ...:   
    ...: @qml.qnode(dev)  
    ...: def circuit(a, b=1):  
    ...:     for idx in range(b):  
    ...:         qml.RX(a[idx], wires=[0])  
    ...:     return qml.expval(qml.PauliX(0))  
    ...:   
    ...: def func():  
    ...:     num_evaluations = 30  
    ...:     for i in range(num_evaluations):  
    ...:         params = np.random.rand(i)  
    ...:         circuit(params, b=i)  
    ...: %timeit func()                                                                                                                                                                                     
90.7 ms ± 381 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

100 evaluations (`num_evaluations = 100`):

```python
In [12]: import pennylane as qml  
    ...: import numpy as np  
    ...:   
    ...: dev = qml.device('default.qubit', wires=1)  
    ...:   
    ...: @qml.qnode(dev)  
    ...: def circuit(a, b=1):  
    ...:     for idx in range(b):  
    ...:         qml.RX(a[idx], wires=[0])  
    ...:     return qml.expval(qml.PauliX(0))  
    ...:   
    ...: def func():  
    ...:     num_evaluations = 100 
    ...:     for i in range(num_evaluations):  
    ...:         params = np.random.rand(i)  
    ...:         circuit(params, b=i)  
    ...: %timeit func()                                                                                                                                                                                     
2.43 s ± 27.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Here going from 30 to 100 evaluations yielded 26.791620728 times of increase in execution time.

--------------------------------

PennyLane 0.9.0-dev with cached `inspect.getfullargspec` (it is called only once per mutable QNode and the first result is used each time)
```python
In [1]: import pennylane as qml  
   ...: import numpy as np  
   ...:   
   ...: dev = qml.device('default.qubit', wires=1)  
   ...:   
   ...: @qml.qnode(dev)  
   ...: def circuit(a, b=1):  
   ...:     for idx in range(b):  
   ...:         qml.RX(a[idx], wires=[0])  
   ...:     return qml.expval(qml.PauliX(0))  
   ...:   
   ...: def func():  
   ...:     num_evaluations = 100 
   ...:     for i in range(num_evaluations):  
   ...:         params = np.random.rand(i)  
   ...:         circuit(params, b=i)  
   ...: %timeit func()                                                                                                                                                                                      
1.81 s ± 178 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Reply:
Thanks @antalszava! So, summing up:

* v0.9 with cached `inspect.getfullargspec`, 100 evals: 1.81 s ± 178 ms

* v0.9 with no caching, 100 evals: 2.43 s ± 27.3 ms

* v0.7 logic, 100 evals: 27.2 ms ± 3.43 ms

This to me is a strong argument that we should revert to v0.7 logic ASAP, and the continue work in a new PR to efficiently return the variable naming feature.

Reply:
Great, thanks for that! :blush:  (sorry for the long and cryptic logs :sweat_smile:)